File: CS50_classes/CS50x/sentimental-credit/credit.py
from cs50 import get_int


def main():
    number = get_card_number()
    check_validity(number)


def get_card_number():
    return get_int("Number: ")


def check_validity(number):
    checksum = calculate_checksum(number)

    if checksum % 10 == 0:
        length = calculate_length(number)

        if isAMEX(number, length):
            print("AMEX")

        elif isMASTERCARD(number, length):
            print("MASTERCARD")

        elif isVISA(number, length):
            print("VISA")
        else:
            print("INVALID")

    else:
        print("INVALID")


def calculate_length(number):
    numDigits = 0

    while number:
        numDigits += 1
        number //= 10

    return numDigits


def calculate_checksum(number):
    odd_sum, even_sum, idx = 0, 0, 0

    odd_sum += number % 10
    number //= 10
    idx += 1

    while number:
        if idx % 2 == 0:
            odd_sum += number % 10

        else:
            tmp = 2 * (number % 10)
            if tmp >= 10:
                even_sum += (tmp % 10) + 1
            else:
                even_sum += tmp

        number //= 10
        idx += 1

    return odd_sum + even_sum


def isAMEX(number, length):
    if length == 15:
        tmp = number // 10000000000000
        if tmp == 34 or tmp == 37:
            return True

    return False


def isMASTERCARD(number, length):
    if length == 16:
        tmp = number // 100000000000000
        if 51 <= tmp <= 55:
            return True

    return False


def isVISA(number, length):
    if length == 13:
        tmp = number // 1000000000000
        if tmp == 4:
            return True

    elif length == 16:
        tmp = number // 1000000000000000
        if tmp == 4:
            return True

    return False


if __name__ == "__main__":
    main()

File: CS50_classes/CS50x/credit/credit.c
#include <cs50.h>
#include <stdio.h>

long get_card_number(void);
void check_validity(long number);
int calculate_length(long number);
int calculate_checksum(long number);

bool isAMEX(long number);
bool isMASTERCARD(long number);
bool isVISA(long number);

int main(void)
{
    long number = get_card_number();
    check_validity(number);
}

long get_card_number(void)
{
    long number = get_long("Number: ");
    return number;
}

void check_validity(long number)
{

    int checksum = calculate_checksum(number);

    if (checksum % 10 == 0)
    {
        if (isAMEX(number))
        {
            printf("AMEX\n");
        }

        else if (isMASTERCARD(number))
        {
            printf("MASTERCARD\n");
        }

        else if (isVISA(number))
        {
            printf("VISA\n");
        }
        else
        {
            printf("INVALID\n");
        }
    }
    else
    {
        printf("INVALID\n");
    }
}

int calculate_length(long number)
{
    int numDigits = 0;

    do
    {
        numDigits++;
        number = number / 10;
    }
    while (number);

    return numDigits;
}

int calculate_checksum(long number)
{
    int odd_sum = 0, even_sum = 0, idx = 0;

    do
    {
        if (idx % 2 == 0)
        {
            odd_sum += number % 10;
        }
        else
        {
            int tmp = 2 * (number % 10);
            if (tmp >= 10)
            {
                even_sum += (tmp % 10) + 1;
            }
            else
            {
                even_sum += tmp;
            }
        }
        number /= 10;
        idx++;
    }
    while (number);

    return odd_sum + even_sum;
}

bool isAMEX(long number)
{
    int length = calculate_length(number);

    if (length == 15)
    {
        int tmp = number / 10000000000000;
        if (tmp == 34 || tmp == 37)
        {
            return true;
        }
    }
    return false;
}

bool isMASTERCARD(long number)
{
    int length = calculate_length(number);
    if (length == 16)
    {
        int tmp = number / 100000000000000;
        if (tmp >= 51 && tmp <= 55)
        {
            return true;
        }
    }
    return false;
}

bool isVISA(long number)
{
    int length = calculate_length(number);
    if (length == 13)
    {
        int tmp = number / 1000000000000;
        if (tmp == 4)
        {
            return true;
        }
    }
    else if (length == 16)
    {
        int tmp = number / 1000000000000000;
        if (tmp == 4)
        {
            return true;
        }
    }
    return false;
}
